### ISSUE TITLE
Wrap Link.__call__

### ISSUE THREAD

**Issue by niboshi** [pfnet/chainer#2748]
Currently users directly override `Link.__call__`.
I propose letting users override, say, `Link.call` instead, and call it from `Link.__call__`. It's just like how `Function.__call__` works.

This will introduce several benefits (At least I can think of):
* Link hooks
* Links will be able to be represented hierarchically in graph visualization. (See #2215)

Some candidate method names: `call`, `__call_link__`, `define`, `define_link`, any other?

Note: This will not break backward compatibility, as existing code (overriding `Link.__call__`) will just work, but with new features (e.g. link hooks) disabled.

**Comment by beam2d** [comment]
It would be better to provide a context manager method or a decorator that inserts the pre/post processes. It will be used like follows.
```py
def __call__(self, x, ...):
    with self.forward_scope():  # run preprocess in __enter__, and postprocess in __exit__
        h = F.relu(...)
        ...
```
or
```py
@Link.forward_method
def __call__(self, x, ...):
    h = F.relu(...)
    ...
```
There are some benefits:

- It is straight-forward to understand the actual interface of the link implementation.
- There is no extra effort needed to support IDE correctly and provide a documentation that reflects the actual interface.
- Users can provide multiple methods that do forward computations for different purposes (e.g. one for computing a loss value and another for inference).

See #2657, too.

**Comment by mitmul** [comment]
"Link hooks" may make ONNX export feature I'm trying to create as Function hooks more easier especially for LSTM-like modules which have multiple Functions inside.

**Comment by beam2d** [comment]
We need to list up possible applications for this feature, then we can start discussing what the interface should be. Any ideas? Here is a list of ideas proposed in the offline discussions (including those that have alread appeared in this issue):
- Graph visualization improvement
- Link hooks
  - ONNX export code injection in Link level (e.g. LSTM)
  - Link-level profiling code injection

**Comment by niboshi** [comment]
I, @bkvogel and @mitmul discussed the interface.
While @beam2d's suggestions above have some advantages, we have a concern that it may be difficult for users (especially those who are novice in Python language) to understand and track the code behind `Link`.
In the alternative (basically it's the same as the one I proposed above), usual user-defined links look just simple (the method name is borrowed from `nn.Module` in PyTorch), and complicated things are required only when multiple forward methods are defined.

Usual user-defined link would look like:
```py
def SomeLink(chainer.Link):
    def forward(self, x):
        h = F.relu(x)
        ....
````

If user wants another forward method (e.g. for `predict`), user needs to mark it in some way. We didn't come up with particular notation, but there may be some possible ways.

```py
# (1) decorator
def SomeLink(chainer.Link):
    def forward(self, x):
        ....
    @Link.forward_method
    def predict(self, x):
        ...

# (2) manual registration
def SomeLink(chainer.Link):
    def __init__(self):
        self.add_forward_method(self.predict)
    def forward(self, x):
        ....
    def predict(self, x):
        ...

# (3) Via class attribute
def SomeLink(chainer.Link):
    extra_forward_methods = ('predict',)
    def forward(self, x):
        ....
    def predict(self, x):
        ....

# (4) Context manager
def SomeLink(chainer.Link):
    def forward(self, x):
        ....
    def predict(self, x):
        with self.forward_scope():
            ....
````

etc.

How do you think? @beam2d

**Comment by tkerola** [comment]
I would like to add another use-case for link hooks: Methods such as spectral normalization (https://arxiv.org/abs/1802.05957) would be very easy to implement in a clean manner, since it could be written as a hook that normalizes the weight matrix before the link is called. Currently, it has to be implemented as `SpectralNormalizationConvolution2D`, `SpectralNormalizationLinear`, `SpectralNormalizationEmbedID`, which causes a lot of unnecessary duplicate code, whereas it would be cleaner to just write a single hook that can be applied to any existing link.

**Comment by corochann** [comment]
I'm not sure if it is relevant or not.
I want to calculate the gradient for link's parameter `W` and `b` for each example in minibatch separately.
And I think this feature can be covered if we can get a hook for `backward` call of function as well.

**Comment by mitmul** [comment]
If there's no objection to the above comment of @niboshi, I'm going to start writing this feature and send a PR soon.

**Comment by niboshi** [comment]
@mitmul
Do you mean `(1) decorator`?
We didn't discuss these options, but I prefer (2) for explicitness in this case, because it will be rarely used.
And I've come to think (4) is not an option because it cannot distinguish different forward methods.

**Comment by mitmul** [comment]
@niboshi I think replacing `__call__` with `forward` in all `Links`s to enable "link hook" feature can be a separate PR from a PR to support multiple forward functions. Can I start to add "link hook" feature first? Then I think we can discuss how to support multiple forward functions.

**Comment by niboshi** [comment]
@mitmul Got it!

**Comment by mitmul** [comment]
Created a PR to add `LinkHook` feature. https://github.com/chainer/chainer/pull/4730
